server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java

```diff
@@ -2,59 +2,189 @@
+            .put(initializationVector)
+            .put(combinedEncoding)
+            .array();
+
+        String s = Base64.getEncoder().encodeToString(combinedEncodingWithIV);
+        return new PrincipalIdentifierToken(s);
+    }
+
+    public static String hex(byte[] bytes) {
+        StringBuilder result = new StringBuilder();
+        for (byte b : bytes) {
+            result.append(String.format("%02x", b));
+        }
+        return result.toString();
     }
 
     /**
      * Decrypt the token and extract Principal
      * @param token the requester identity token, should not be null
      * @return Principal
+     * @throws InvalidAlgorithmParameterException
+     * @throws InvalidKeyException
+     * @throws BadPaddingException
+     * @throws IllegalBlockSizeException
+     * @throws NoSuchPaddingException
+     * @throws NoSuchAlgorithmException
      *
      * @opensearch.internal
-     *
-     * This method contains a placeholder implementation.
-     * More concrete implementation will be covered in https://github.com/opensearch-project/OpenSearch/issues/4485
      */
-    public Principal extractPrincipal(PrincipalIdentifierToken token) throws IllegalArgumentException {
-        // check is token is valid, we don't do anything if it is valid
-        // else we re-throw the thrown exception
-        validateToken(token);
-
-        String[] parts = token.getToken().split(":");
-        final String principalName = parts[0];
-        return () -> principalName;
+    public String extractPrincipal(PrincipalIdentifierToken token, SecretKey secretKey) throws IllegalArgumentException,
+        InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, NoSuchAlgorithmException,
+        NoSuchPaddingException {
+
+        String tokenString = token.getToken();
+        byte[] tokenBytes = Base64.getDecoder().decode(tokenString);
+
+        ByteBuffer bb = ByteBuffer.wrap(tokenBytes);
+
+        byte[] iv = new byte[INITIALIZATION_VECTOR_SIZE_BYTES];
+        bb.get(iv);
+
+        byte[] cipherText = new byte[bb.remaining()];
+        bb.get(cipherText);
+
+        Cipher decryptionCipher = Cipher.getInstance("AES/GCM/NoPadding");
+        GCMParameterSpec spec = new GCMParameterSpec(TAG_LENGTH_BITS, iv);
+
+        try {
+            decryptionCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+        } catch (InvalidKeyException ex) {
+            throw new InvalidKeyException(INVALID_KEY_MESSAGE);
+        }
+
+        byte[] combinedEncoding;
+        try {
+            combinedEncoding = decryptionCipher.doFinal(cipherText);
+        } catch (AEADBadTagException ex) {
+            throw new AEADBadTagException(INVALID_TAG_MESSAGE);
+        }
+
+        String decodedPrincipal = new String(combinedEncoding, StandardCharsets.UTF_8).replace(this.extensionUniqueId, "");
```

Comment:
this is causing forbiddenApis to fail. All instances where byte[] is converted to String will throw this